### PR TITLE
Add support for writing CSV data sequentially

### LINKED ISSUE #13: Allow for writing a stream of data to a CSV file
At present, all of the data is required to be in memory (e.g. as a ``DataFrame``) before it can be written out to a CSV file. This is not a problem for small datasets, but for large ones -- or ones that are acquired over a long time period -- it is important to be able to write data as you go along (partly because of memory constraints, but also because e.g. the power might fail).

This is the sort of interface I have in mind:
```py
with csvy.StreamWriter(filename, columns=['Time', 'CH0', 'CH1'], metadata=metadata) as writer:
    while something():
        writer.write_row(time, *data)
```

### PR DESCRIPTION
**Pull request by alexdewar**
This PR adds a new `Writer` class, which allows the user to write data sequentially. The problem with the existing code is that all the data must reside in memory before it can be written with write(). For large datasets or those that are being acquired in real time (e.g. from a sensor), this approach doesn't work.

The Writer class provides much the same interface as the writer objects returned from csv.writer. There is also a context manager for automatically closing the file handle.

Closes #13.

### CODE REVIEW

**Comment by alexdewar** [comment]
This PR also needs updating after #24 is merged.

**Review by dalonsoa — CHANGES_REQUESTED** [review]
Looks great! The only request really is to add tests for all the added functionality (you can see what's missing in the inline comments of the Codecov report).

**Review by dalonsoa — APPROVED** [review]
Looks good! Just a minor comment.

**Comment by dalonsoa** [comment]
@all-contributors please add @alexdewar  for  test and code

File: csvy/writers.py [hunk 1]
```diff
@@ -42,28 +44,89 @@ def write(
...
+        if not yaml_options:
+            yaml_options = {}
+
+        # Newline must be "" as per csv.writer's documentation
+        self._file = Path(filename).open("w", newline="")
+        write_header(self._file, header, comment, **yaml_options)
+
+        self._writer = csv.writer(self._file, **csv_options)
+
+    def __enter__(self) -> "Writer":
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply csvy/writers.py:83]: Ahh that's interesting. TIL!

**Pull request merged**